### PR TITLE
#1417 Add documentation about the builder support

### LINKED ISSUE #1417: Add documentation about the Builders support
How a builder is detect?
How are builder create method and build method detected?

### PR DESCRIPTION
**Pull request by filiphr**
Fixes #1417

### CODE REVIEW

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 1]
```diff
@@ -594,6 +611,106 @@ You can find the complete example in the
...
 
+[[mapping-with-builders]]
+=== Using builders
+
+MapStruct also supports mapping of immutable types via builders.
+When performing a mapping MapStruct checks if there is a builder for the type being mapped.
+This is done via the `BuilderProvider` SPI.
+If a Builder exists for a certain type, than that builder will be used for the mappings.
+
+The default implementation of the `BuilderProvider` requires the following for a type to have a builder:
```
Comment (sjaakd) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: This line does not run smoothly. .

Comment (sjaakd) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: So: `The default implementation of the `BuilderProvider` assumes the following:

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 2]
```diff
@@ -594,6 +611,106 @@ You can find the complete example in the
...
+
+* The type (`Person`) has a parameterless public static method (builder creation method) that returns a type (`PersonBuilder`) different than the one that needs building
```
Comment (sjaakd) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: Split in line + example: The type has a parameterless public static builder creation method that returns a builder of the type. So `Person` has a for example a method `createPerson` that returns `PersonBuilder`.

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 3]
```diff
@@ -594,6 +611,106 @@ You can find the complete example in the
...
+* The found builder type (`PersonBuilder`) has a parameterless public method (build method) that returns the type being build (`Person`)
```
Comment (sjaakd) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: split: The builder type has a parameterless public method (build method) that returns the type being build. In our example: `PersonBuilder` has a method returning `Person`.

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 4]
```diff
@@ -594,6 +611,106 @@ You can find the complete example in the
...
+
+If such type is found then MapStruct will use that type to perform the mapping to (i.e. it will look for setters into that type).
+If no factory method exists for instantiating the builder type the builder creation method would be used.
```
Comment (sjaakd) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: factory method is new.. you just called it a method before. Align.

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 5]
```diff
@@ -594,6 +611,106 @@ You can find the complete example in the
...
+To finish the mapping MapStruct will invoke the build method of the builder.
```
Comment (sjaakd) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: To finish the mapping MapStruct generates code that will invoke the build method of the builder.

**Comment by sjaakd** [comment]
> As if there is a factory method (which we use in the documentation) for the PersonBuilder. Then that method would be used instead of the builder creation method

Ok. Didn't understand it like this.. Perhaps good to refer to chapter on factory methods. Btw; is this valid for automatically selected factory methods and/or annotated ones?

**Comment by filiphr** [comment]
>Btw; is this valid for automatically selected factory methods and/or annotated ones?

We basically delegate the builder type to the resolver. I'll write a test for it as well.

**Comment by filiphr** [comment]
@sjaakd I've updated the documentation. I noticed that Factories don't really work for Builders. I will do the fix in a new PR

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc (hunk 4 above)
Comment (sjaakd) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: refer to chapter 9...(ObjectFactoryies)

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 6]
```diff
@@ -627,9 +627,14 @@ So for example `Person` has a public static method that returns `PersonBuilder`.
 In our example `PersonBuilder` has a method returning `Person`.
 
 If such type is found then MapStruct will use that type to perform the mapping to (i.e. it will look for setters into that type).
-If no factory method exists for instantiating the builder type the builder creation method would be used.
 To finish the mapping MapStruct generates code that will invoke the build method of the builder.
 
+[NOTE]
+======
+The <<object-factories>> lifecylce is also considered for the builder type.
```
Comment (sjaakd) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: the term " lifecycle"  is confusing..

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc (hunk 4 above)
Comment (filiphr) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: I've removed that line actually

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc (hunk 6 above)
Comment (filiphr) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: That's true. What about just:

>The <<object-factories>> are also considered...

Comment (sjaakd) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: 👍

**Pull request merged**